src/NuGet.Core/NuGet.DependencyResolver.Core/GraphModel/GraphNode.cs

```diff
@@ -21,6 +24,25 @@ public GraphNode(LibraryRange key)
         public IList<GraphNode<TItem>> InnerNodes { get; set; }
         public Disposition Disposition { get; set; }
 
+        /// <summary>
+        /// Used in case that a node is removed from its outernode and needs to keep reference of its parents.
+        /// </summary>
+        internal BlockingCollection<GraphNode<TItem>> ParentNodes { get; }
```

Comment:
Why are we using a blocking collection? Is it solely because of IsAddingCompleted?

Reply:
Yes, it was because the IsAddingCompleted. However this was an extra and probably not necessary caution because the iterations of the collections is executed after all the parents are added. Running performance tests iterating over BlockingCollection is more expensive so changed to IList.